File: pipeline/pjsr/corona-drift.js
#engine v8   // must sit in the first few lines: PixInsight only scans a
             // short prologue for it, and a header comment long enough to
             // push it past ~line 20 leaves the script in ES5, where let
             // and arrow functions are a load error that -r mode discards
             // silently with exit 0 and no log.
/*
 * corona-drift.js — measure how fast the Moon slides across the corona.
 *
 * The Moon and the corona are not the same reference frame. The corona is fixed
 * to the Sun; the Moon crosses it at roughly its own synodic rate. Over one 60 s
 * capture that is a handful of pixels, but across the whole exposure ladder it is
 * tens of pixels — enough to double the streamers in an HDR merge, or to double
 * the limb, depending on which feature the registration happened to lock onto.
 *
 * Rather than trust an ephemeris and an assumed image scale, this measures the
 * rate from the data: take two stacks of the SAME exposure level at different
 * times, put them in a common Moon-centred frame, mask out everything but the
 * outer corona, and cross-correlate. Whatever shift remains is the Moon-vs-Sun
 * slip over the known interval.
 *
 * The correlation runs on the radially flattened image. Without that, the profile
 * itself dominates the correlation and every candidate shift near zero scores
 * about the same, because a smooth radial gradient correlates with itself.
 *
 *   -r="...corona-drift.js,<config.json>,<logPath>"
 *
 * Config: { "a": {"path":...,"cx":...,"cy":...,"r":...,"t":secs},
 *           "b": {...}, "out": "drift.json" }
 */


// Annulus used for the correlation, in units of the Moon's radius. The inner
// bound clears the limb and the chromosphere. The outer bound is deliberately
// close in: a first attempt reaching out to 3.2 R measured a drift three times
// the physical ceiling, because that far out the corona is faint and the
// correlation locked onto the sensor's own flat-field signature instead, which
// does not move between frames and so reported the Moon's full frame motion as
// if it were differential. Structure this close in is unambiguously corona.
const R_INNER = 1.10;
const R_OUTER = 2.00;

// Physical ceiling on the differential rate. The Moon moves relative to the Sun
// at the synodic rate; converting through the measured Moon radius turns that
// into pixels per second without needing to know the focal length. Anything
// above this is a failed correlation, not a fast Moon - the topocentric rate
// during totality is lower still, so the margin is generous.
const SYNODIC_ARCSEC_PER_S = 0.51;
const MOON_RADIUS_ARCSEC = 990.0;
const DRIFT_SAFETY = 1.4;

// Coarse search is done on an 8x-decimated image, so this window covers
// +/-(SEARCH_COARSE*8) full-resolution pixels.
const SEARCH_COARSE = 14;
const DECIM = 8;
const SEARCH_FINE = 6;

function main()
{
   let cfgPath = jsArguments[0];
   let logPath = jsArguments[1];

   let logFile = new File;
   logFile.createForWriting( logPath );
   function log( s )
   {
      logFile.outTextLn( String( s ) );
      logFile.flush();
      console.writeln( String( s ) );
      console.flush();
   }

   try
   {
      let t0 = Date.now();
      // File.readTextFile does not exist; a missing PJSR API kills the script
      // silently with exit 0. utf8ToString on a ByteArray is the working idiom.
      let cfg = JSON.parse( File.readFile( cfgPath ).utf8ToString().replace( /^\uFEFF/, "" ) );
      let dt = cfg.b.t - cfg.a.t;
      if ( !(Math.abs( dt ) > 1 ) )
         throw new Error( "need a meaningful time baseline, got dt=" + dt + "s" );

      let A = ecLoadPlane( cfg.a, log );
      let B = ecLoadPlane( cfg.b, log );
      if ( A.W != B.W || A.H != B.H )
         throw new Error( "geometry mismatch between the two stacks" );
      let W = A.W, H = A.H;

      // Common Moon-centred frame: shift B so its Moon sits on A's Moon. Any
      // residual corona shift after this is the differential motion. The offset
      // comes from corona-register.js when available, since differencing two
      // independent Moon fits is exactly what proved unreliable.
      let mx, my;
      if ( cfg.b.dx !== undefined && cfg.a.dx !== undefined )
      {
         mx = cfg.b.dx - cfg.a.dx;
         my = cfg.b.dy - cfg.a.dy;
      }
      else
      {
         mx = cfg.a.cx - cfg.b.cx;
         my = cfg.a.cy - cfg.b.cy;
      }
      let bAligned = translateImage( B.lum, W, H, mx, my );
      log( "  moon-registered B onto A: (" + mx.toFixed( 2 ) + ", "
           + my.toFixed( 2 ) + ") px" );

      let rIn = cfg.a.r*R_INNER, rOut = cfg.a.r*R_OUTER;
      let fa = flattenAnnulus( A.lum, W, H, cfg.a.cx, cfg.a.cy, rIn, rOut );
      let fb = flattenAnnulus( bAligned, W, H, cfg.a.cx, cfg.a.cy, rIn, rOut );

      // Coarse pass on a decimated grid, then refine at full resolution.
      let dA = decimate( fa.img, fa.mask, W, H, DECIM );
      let dB = decimate( fb.img, fb.mask, W, H, DECIM );
      let coarse = search( dA, dB, dA.w, dA.h, SEARCH_COARSE );
      log( "  coarse best (" + (coarse.dx*DECIM) + ", " + (coarse.dy*DECIM)
           + ") ncc=" + coarse.score.toFixed( 5 ) );

      let fine = searchAround( fa, fb, W, H,
                               coarse.dx*DECIM, coarse.dy*DECIM, SEARCH_FINE );
      log( "  refined best (" + fine.dx.toFixed( 2 ) + ", " + fine.dy.toFixed( 2 )
           + ") ncc=" + fine.score.toFixed( 5 ) );

      // fine.* is how far B's corona sits from A's after Moon registration, so
      // the corona moves by -fine per dt in the Moon frame.
      let vx = -fine.dx/dt, vy = -fine.dy/dt;
      let speed = Math.sqrt( vx*vx + vy*vy );
      log( "  baseline " + dt.toFixed( 1 ) + " s" );
      log( "  differential drift " + speed.toFixed( 4 ) + " px/s  ("
           + vx.toFixed( 4 ) + ", " + vy.toFixed( 4 ) + ")" );
      log( "  => " + (speed*60).toFixed( 1 ) + " px per 60 s capture" );

      // Physics gate. A measured rate above the synodic ceiling means the
      // correlation found something that is not corona; report it and hand back
      // zero so the caller falls back to rigid registration rather than shifting
      // every level by a fabricated amount.
      let ceiling = DRIFT_SAFETY*SYNODIC_ARCSEC_PER_S*cfg.a.r/MOON_RADIUS_ARCSEC;
      let accepted = speed <= ceiling;
      log( "  ceiling " + ceiling.toFixed( 4 ) + " px/s (moon r=" + cfg.a.r.toFixed( 1 )
           + " px) -> " + (accepted ? "accepted" : "REJECTED") );
      if ( !accepted )
      {
         log( "  *** measured rate exceeds what the Moon can do; reporting zero drift" );
         vx = 0; vy = 0;
      }

      let out = new File;
      out.createForWriting( cfg.out );
      out.outTextLn( JSON.stringify( {
         dtSeconds: dt,
         shiftPx: { dx: fine.dx, dy: fine.dy },
         driftPxPerSec: { x: vx, y: vy, speed: Math.sqrt( vx*vx + vy*vy ) },
         measuredSpeed: speed,
         ceiling: ceiling,
         accepted: accepted,
         ncc: fine.score,
         annulus: { rInner: rIn, rOuter: rOut },
      } ) );
      out.close();

      log( "  wrote " + cfg.out + " in " + ((Date.now() - t0)/1000).toFixed( 0 ) + " s" );
      log( "=== DRIFT OK ===" );
   }
   catch ( e )
   {
      log( "*** DRIFT FAILED: " + e.toString() );
      if ( e.stack )
         log( String( e.stack ) );
   }
   logFile.close();
}

function ecLoadPlane( spec, log )
{
   let wins = ImageWindow.open( spec.path );
   let win = wins[0];
   for ( let i = 1; i < wins.length; ++i )
      wins[i].forceClose();
   let img = win.mainView.image;
   let W = img.width, H = img.height;
   let lum = new Float32Array( W*H );
   img.getSamples( lum, new Rect( 0, 0, W, H ), img.numberOfChannels > 1 ? 1 : 0 );
   win.forceClose();
   log( "  loaded " + spec.path.substring( spec.path.lastIndexOf( "/" ) + 1 ) + " " + W + "x" + H );
   return { lum: lum, W: W, H: H };
}

/*
 * Divide out the radial median profile inside the annulus and return the result
 * plus a validity mask. Flattening is what gives the correlation something to
 * lock onto: the streamers, not the falloff.
 */
function flattenAnnulus( g, W, H, cx, cy, rIn, rOut )
{
   let n = W*H;
   let rmaxI = Math.ceil( rOut );
   let sum = new Float64Array( rmaxI + 1 );
   let cnt = new Int32Array( rmaxI + 1 );
   for ( let y = 0; y < H; ++y )
   {
      let dy = y - cy, row = y*W;
      for ( let x = 0; x < W; ++x )
      {
         let dx = x - cx;
         let d = Math.sqrt( dx*dx + dy*dy );
         if ( d < rIn || d > rOut ) continue;
         let r = Math.round( d );
         sum[r] += g[row + x];
         cnt[r]++;
      }
   }
   let prof = new Float64Array( rmaxI + 1 );
   for ( let r = 0; r <= rmaxI; ++r )
      prof[r] = cnt[r] > 0 ? sum[r]/cnt[r] : 0;

   let img = new Float32Array( n );
   let mask = new Uint8Array( n );
   for ( let y = 0; y < H; ++y )
   {
      let dy = y - cy, row = y*W;
      for ( let x = 0; x < W; ++x )
      {
         let dx = x - cx;
         let d = Math.sqrt( dx*dx + dy*dy );
         if ( d < rIn || d > rOut ) continue;
         let r = Math.round( d );
         let p = prof[r];
         if ( !(p > 0) ) continue;
         img[row + x] = g[row + x]/p - 1;   // fractional excess over the profile
         mask[row + x] = 1;
      }
   }
   return { img: img, mask: mask };
}

function decimate( img, mask, W, H, k )
{
   let w = Math.floor( W/k ), h = Math.floor( H/k );
   let oi = new Float32Array( w*h );
   let om = new Uint8Array( w*h );
   for ( let y = 0; y < h; ++y )
      for ( let x = 0; x < w; ++x )
      {
         let s = 0, c = 0;
         for ( let j = 0; j < k; ++j )
         {
            let row = (y*k + j)*W + x*k;
            for ( let i = 0; i < k; ++i )
               if ( mask[row + i] ) { s += img[row + i]; c++; }
         }
         if ( c > (k*k >> 1) ) { oi[y*w + x] = s/c; om[y*w + x] = 1; }
      }
   return { img: oi, mask: om, w: w, h: h };
}

/* Normalized cross-correlation over integer shifts, masked. */
function ncc( a, b, W, H, dx, dy )
{
   let sa = 0, sb = 0, saa = 0, sbb = 0, sab = 0, n = 0;
   for ( let y = 0; y < H; ++y )
   {
      let sy = y + dy;
      if ( sy < 0 || sy >= H ) continue;
      let ra = y*W, rb = sy*W;
      for ( let x = 0; x < W; ++x )
      {
         let sx = x + dx;
         if ( sx < 0 || sx >= W ) continue;
         if ( !a.mask[ra + x] || !b.mask[rb + sx] ) continue;
         let u = a.img[ra + x], v = b.img[rb + sx];
         sa += u; sb += v; saa += u*u; sbb += v*v; sab += u*v; n++;
      }
   }
   if ( n < 500 ) return -2;
   let ma = sa/n, mb = sb/n;
   let ca = saa - n*ma*ma, cb = sbb - n*mb*mb;
   if ( ca <= 0 || cb <= 0 ) return -2;
   return (sab - n*ma*mb)/Math.sqrt( ca*cb );
}

function search( a, b, W, H, win )
{
   let best = { dx: 0, dy: 0, score: -2 };
   for ( let dy = -win; dy <= win; ++dy )
      for ( let dx = -win; dx <= win; ++dx )
      {
         let s = ncc( a, b, W, H, dx, dy );
         if ( s > best.score ) best = { dx: dx, dy: dy, score: s };
      }
   return best;
}

function searchAround( a, b, W, H, cx, cy, win )
{
   let best = { dx: cx, dy: cy, score: -2 };
   for ( let dy = cy - win; dy <= cy + win; ++dy )
      for ( let dx = cx - win; dx <= cx + win; ++dx )
      {
         let s = ncc( a, b, W, H, dx, dy );
         if ( s > best.score ) best = { dx: dx, dy: dy, score: s };
      }
   return best;
}

function translateImage( src, W, H, dx, dy )
{
   let dst = new Float32Array( W*H );
   for ( let y = 0; y < H; ++y )
   {
      let sy = y - dy;
      let y0 = Math.floor( sy ), fy = sy - y0, y1 = y0 + 1;
      if ( y1 < 0 || y0 >= H ) continue;
      if ( y0 < 0 ) { y0 = 0; fy = 0; }
      if ( y1 >= H ) y1 = H - 1;
      let r0 = y0*W, r1 = y1*W, drow = y*W;
      for ( let x = 0; x < W; ++x )
      {
         let sx = x - dx;
         let x0 = Math.floor( sx ), fx = sx - x0, x1 = x0 + 1;
         if ( x1 < 0 || x0 >= W ) continue;
         if ( x0 < 0 ) { x0 = 0; fx = 0; }
         if ( x1 >= W ) x1 = W - 1;
         let u = src[r0 + x0] + (src[r0 + x1] - src[r0 + x0])*fx;
         let v = src[r1 + x0] + (src[r1 + x1] - src[r1 + x0])*fx;
         dst[drow + x] = u + (v - u)*fy;
      }
   }
   return dst;
}

main();

File: pipeline/ecl/corona_drift.py
"""Measure how fast the Moon slides across the corona — port of corona-drift.js.

The Moon and the corona are not the same reference frame. The corona is fixed to
the Sun; the Moon crosses it at roughly its own synodic rate. Over one 60 s
capture that is a handful of pixels, but across the whole exposure ladder it is
tens of pixels — enough to double the streamers in an HDR merge, or to double the
limb, depending on which feature the registration happened to lock onto.

Rather than trust an ephemeris and an assumed image scale, this measures the rate
from the data: take two stacks of the SAME exposure level at different times, put
them in a common Moon-centred frame, mask out everything but the outer corona,
and cross-correlate. Whatever shift remains is the Moon-vs-Sun slip over the
known interval.

The correlation runs on the radially FLATTENED image. Without that the profile
itself dominates and every candidate shift near zero scores about the same,
because a smooth radial gradient correlates with itself.
"""

import argparse
import json
import math
import os
import time

import numpy as np

from .imgio import read_xisf

# Annulus for the correlation, in units of the Moon's radius. The inner bound
# clears the limb and the chromosphere. The outer bound is deliberately close in:
# a first attempt reaching to 3.2 R measured a drift three times the physical
# ceiling, because that far out the corona is faint and the correlation locked
# onto the sensor's own flat-field signature, which does not move between frames
# and so reported the Moon's full frame motion as if it were differential.
R_INNER = 1.10
R_OUTER = 2.00

# Physical ceiling on the differential rate. Converting the synodic rate through
# the MEASURED Moon radius gives pixels per second without needing the focal
# length. Anything above this is a failed correlation, not a fast Moon.
SYNODIC_ARCSEC_PER_S = 0.51
MOON_RADIUS_ARCSEC = 990.0
DRIFT_SAFETY = 1.4

SEARCH_COARSE = 14
DECIM = 8
SEARCH_FINE = 6
MIN_OVERLAP = 500

__all__ = ["measure_drift", "flatten_annulus", "pick_drift_pair"]


def flatten_annulus(g, cx, cy, r_in, r_out):
    """Divide out the azimuthally averaged radial profile inside an annulus.

    Returns the fractional excess over the profile, plus the annulus mask. What
    is left is what the correlation should lock onto: the streamers, not the
    falloff.
    """
    H, W = g.shape
    yy, xx = np.mgrid[0:H, 0:W]
    d = np.hypot(xx - cx, yy - cy)
    inside = (d >= r_in) & (d <= r_out)

    r_max = int(math.ceil(r_out))
    ri = np.floor(d + 0.5).astype(np.int64)          # JS Math.round
    np.clip(ri, 0, r_max, out=ri)
    sel = ri[inside]
    cnt = np.bincount(sel, minlength=r_max + 1)
    tot = np.bincount(sel, weights=g[inside].astype(np.float64),
                      minlength=r_max + 1)
    prof = np.where(cnt > 0, tot / np.maximum(cnt, 1), 0.0)

    p = prof[ri]
    ok = inside & (p > 0)
    img = np.zeros_like(g, dtype=np.float32)
    img[ok] = (g[ok] / p[ok] - 1.0).astype(np.float32)
    return img, ok


def decimate(img, mask, k):
    """Block mean over masked pixels, keeping a cell only if it is mostly valid."""
    h, w = img.shape
    hh, ww = h // k, w // k
    im = img[:hh * k, :ww * k].reshape(hh, k, ww, k)
    mk = mask[:hh * k, :ww * k].reshape(hh, k, ww, k)
    c = mk.sum(axis=(1, 3))
    s = np.where(mk, im, 0).sum(axis=(1, 3))
    keep = c > (k * k >> 1)
    out = np.zeros((hh, ww), np.float32)
    np.divide(s, np.maximum(c, 1), out=out, where=keep)
    return out, keep


def ncc(a_img, a_mask, b_img, b_mask, dx, dy):
    """Masked normalized cross-correlation of b shifted by (dx, dy) against a."""
    h, w = a_img.shape
    y0, y1 = max(0, -dy), min(h, h - dy)
    x0, x1 = max(0, -dx), min(w, w - dx)
    if y1 <= y0 or x1 <= x0:
        return -2.0
    A = a_img[y0:y1, x0:x1]
    B = b_img[y0 + dy:y1 + dy, x0 + dx:x1 + dx]
    M = a_mask[y0:y1, x0:x1] & b_mask[y0 + dy:y1 + dy, x0 + dx:x1 + dx]
    n = int(M.sum())
    if n < MIN_OVERLAP:
        return -2.0
    u = A[M].astype(np.float64)
    v = B[M].astype(np.float64)
    ma, mb = u.mean(), v.mean()
    ca = float((u * u).sum()) - n * ma * ma
    cb = float((v * v).sum()) - n * mb * mb
    if ca <= 0 or cb <= 0:
        return -2.0
    return float((float((u * v).sum()) - n * ma * mb) / math.sqrt(ca * cb))


def _search(a, am, b, bm, cx, cy, win):
    best = (cx, cy, -2.0)
    for dy in range(cy - win, cy + win + 1):
        for dx in range(cx - win, cx + win + 1):
            s = ncc(a, am, b, bm, dx, dy)
            if s > best[2]:
                best = (dx, dy, s)
    return best


def translate_image(src, dx, dy):
    """Bilinear translate, leaving zero outside the source.

    Replicating the edge pixel instead would smear a bright border inward and it
    would pass the signal test as if it were data.
    """
    H, W = src.shape
    ys = np.arange(H) - dy
    xs = np.arange(W) - dx
    y0 = np.floor(ys).astype(np.int64)
    x0 = np.floor(xs).astype(np.int64)
    fy = (ys - y0).astype(np.float32)
    fx = (xs - x0).astype(np.float32)
    y1, x1 = y0 + 1, x0 + 1

    vy = (y1 >= 0) & (y0 < H)
    vx = (x1 >= 0) & (x0 < W)
    # Clamp the fractional weight where the pair straddles the border, matching
    # the original's `if (y0 < 0) { y0 = 0; fy = 0; }`.
    fy = np.where(y0 < 0, 0.0, fy)
    fx = np.where(x0 < 0, 0.0, fx)
    y0c, y1c = np.clip(y0, 0, H - 1), np.clip(y1, 0, H - 1)
    x0c, x1c = np.clip(x0, 0, W - 1), np.clip(x1, 0, W - 1)

    dst = np.zeros_like(src)
    Y0, Y1 = y0c[vy][:, None], y1c[vy][:, None]
    X0, X1 = x0c[vx][None, :], x1c[vx][None, :]
    FY, FX = fy[vy][:, None], fx[vx][None, :]
    u = src[Y0, X0] + (src[Y0, X1] - src[Y0, X0]) * FX
    v = src[Y1, X0] + (src[Y1, X1] - src[Y1, X0]) * FX
    dst[np.ix_(np.nonzero(vy)[0], np.nonzero(vx)[0])] = u + (v - u) * FY
    return dst


def _plane(path):
    img = read_xisf(path)
    return img[:, :, 1] if img.ndim == 3 else img


def measure_drift(a, b, log=print):
    """`a` and `b` are dicts of {path, cx, cy, r, t} plus optional {dx, dy}."""
    t0 = time.time()
    dt = b["t"] - a["t"]
    if not abs(dt) > 1:
        raise ValueError(f"need a meaningful time baseline, got dt={dt}s")

    A, B = _plane(a["path"]), _plane(b["path"])
    if A.shape != B.shape:
        raise ValueError("geometry mismatch between the two stacks")

    # Common Moon-centred frame: shift B so its Moon sits on A's. Any residual
    # corona shift after this is the differential motion. The offset comes from
    # the registration when available, since differencing two independent Moon
    # fits is exactly what proved unreliable.
    if a.get("dx") is not None and b.get("dx") is not None:
        mx, my = b["dx"] - a["dx"], b["dy"] - a["dy"]
    else:
        mx, my = a["cx"] - b["cx"], a["cy"] - b["cy"]
    b_aligned = translate_image(B, mx, my)
    log(f"  moon-registered B onto A: ({mx:.2f}, {my:.2f}) px")

    r_in, r_out = a["r"] * R_INNER, a["r"] * R_OUTER
    fa, ma = flatten_annulus(A, a["cx"], a["cy"], r_in, r_out)
    fb, mb = flatten_annulus(b_aligned, a["cx"], a["cy"], r_in, r_out)

    da, dam = decimate(fa, ma, DECIM)
    db, dbm = decimate(fb, mb, DECIM)
    cdx, cdy, cs = _search(da, dam, db, dbm, 0, 0, SEARCH_COARSE)
    log(f"  coarse best ({cdx * DECIM}, {cdy * DECIM}) ncc={cs:.5f}")

    fdx, fdy, fs = _search(fa, ma, fb, mb, cdx * DECIM, cdy * DECIM, SEARCH_FINE)
    log(f"  refined best ({fdx:.2f}, {fdy:.2f}) ncc={fs:.5f}")

    # The refined shift is how far B's corona sits from A's after Moon
    # registration, so the corona moves by -shift per dt in the Moon frame.
    vx, vy = -fdx / dt, -fdy / dt
    speed = math.hypot(vx, vy)
    log(f"  baseline {dt:.1f} s")
    log(f"  differential drift {speed:.4f} px/s  ({vx:.4f}, {vy:.4f})")
    log(f"  => {speed * 60:.1f} px per 60 s capture")

    # Physics gate. A rate above the synodic ceiling means the correlation found
    # something that is not corona; hand back zero so the caller falls back to
    # rigid registration rather than shifting every level by a fabricated amount.
    ceiling = DRIFT_SAFETY * SYNODIC_ARCSEC_PER_S * a["r"] / MOON_RADIUS_ARCSEC
    accepted = speed <= ceiling
    log(f"  ceiling {ceiling:.4f} px/s (moon r={a['r']:.1f} px) -> "
        f"{'accepted' if accepted else 'REJECTED'}")
    if not accepted:
        log("  *** measured rate exceeds what the Moon can do; reporting zero drift")
        vx = vy = 0.0

    log(f"  measured in {time.time() - t0:.0f} s")
    return {
        "dtSeconds": dt,
        "shiftPx": {"dx": fdx, "dy": fdy},
        "driftPxPerSec": {"x": vx, "y": vy, "speed": math.hypot(vx, vy)},
        "measuredSpeed": speed,
        "ceiling": ceiling,
        "accepted": accepted,
        "ncc": fs,
        "annulus": {"rInner": r_in, "rOuter": r_out},
    }


def pick_drift_pair(images):
    """Two stacks of the same exposure level with the longest time baseline."""
    by_level = {}
    for im in images:
        by_level.setdefault(im["level"], []).append(im)
    best = None
    for group in by_level.values():
        if len(group) < 2:
            continue
        s = sorted(group, key=lambda i: i["t"])
        dt = abs(s[-1]["t"] - s[0]["t"])
        if best is None or dt > best["dt"]:
            best = {"a": s[0], "b": s[-1], "dt": dt}
    return best


def main(argv=None):
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("--config", required=True,
                    help="corona_drift.json with {a, b, out}")
    args = ap.parse_args(argv)
    with open(args.config, encoding="utf-8-sig") as f:
        cfg = json.load(f)
    out = measure_drift(cfg["a"], cfg["b"])
    os.makedirs(os.path.dirname(os.path.abspath(cfg["out"])), exist_ok=True)
    with open(cfg["out"], "w", encoding="utf-8") as f:
        json.dump(out, f)
    print(f"wrote {cfg['out']}")


if __name__ == "__main__":
    main()
